cpp/include/cuvs/neighbors/ivf_flat.hpp

```diff
@@ -0,0 +1,331 @@
+  {
+    return raft_index_->lists();
+  }
+
+  auto get_raft_index() const -> const raft::neighbors::ivf_flat::index<T, IdxT>*
+  {
+    return raft_index_.get();
+  }
+  auto get_raft_index() -> raft::neighbors::ivf_flat::index<T, IdxT>*
+  {
+    return raft_index_.get();
+  }
+
+ private:
+  std::unique_ptr<raft::neighbors::ivf_flat::index<T, IdxT>> raft_index_;
+};
+
+#define CUVS_IVF_FLAT(T, IdxT)                                                                \
+  auto build(raft::resources const& handle,                                                   \
+             const cuvs::neighbors::ivf_flat::index_params& params,                           \
+             raft::device_matrix_view<const T, IdxT, raft::row_major> dataset)                \
+    -> cuvs::neighbors::ivf_flat::index<T, IdxT>;                                             \
+                                                                                              \
+  void build(raft::resources const& handle,                                                   \
+             const cuvs::neighbors::ivf_flat::index_params& params,                           \
+             raft::device_matrix_view<const T, IdxT, raft::row_major> dataset,                \
+             cuvs::neighbors::ivf_flat::index<T, IdxT>& idx);                                 \
+                                                                                              \
+  auto extend(raft::resources const& handle,                                                  \
+              raft::device_matrix_view<const T, IdxT, raft::row_major> new_vectors,           \
+              std::optional<raft::device_vector_view<const IdxT, IdxT>> new_indices,          \
+              const cuvs::neighbors::ivf_flat::index<T, IdxT>& orig_index)                    \
+    -> cuvs::neighbors::ivf_flat::index<T, IdxT>;                                             \
+                                                                                              \
+  void extend(raft::resources const& handle,                                                  \
+              raft::device_matrix_view<const T, IdxT, raft::row_major> new_vectors,           \
+              std::optional<raft::device_vector_view<const IdxT, IdxT>> new_indices,          \
+              cuvs::neighbors::ivf_flat::index<T, IdxT>* idx);                                \
+                                                                                              \
+  void search(raft::resources const& handle,                                                  \
+              const cuvs::neighbors::ivf_flat::search_params& params,                         \
+              cuvs::neighbors::ivf_flat::index<T, IdxT>& index,                               \
+              raft::device_matrix_view<const T, IdxT, raft::row_major> queries,               \
+              raft::device_matrix_view<IdxT, IdxT, raft::row_major> neighbors,                \
```

Comment:
Thanks @viclafargue for this PR! I think it is important to make a distinction between index type that is used to return neighbor indices (`IdxT`) and the index type used to calculate offsets in mdspans (lets call it `MatIdxT`).  I think it would be good to have MatIdxT consistently `int64_t`, while IdxT could remain a template parameter. that we specialize for int64_t.

Reply:
Thanks for the review! Makes sense. Can totally do the change, however it won't work when `IdxT != MatIdxT`. An update of the RAFT runtime would be necessary for this to work.